// ng-in-viewport/esm5/lib/in-viewport.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Directive, ElementRef, EventEmitter, Inject, Input, Output, PLATFORM_ID } from '@angular/core';
import { isPlatformBrowser } from '@angular/common';
import { Subscription } from 'rxjs';
import { filter } from 'rxjs/operators';
import { InViewportConfig } from './in-viewport-config';
import { InViewportService } from './in-viewport.service';
export var /** @type {?} */ InViewportMetadata = Symbol('InViewportMetadata');
var InViewportDirective = /** @class */ (function () {
    function InViewportDirective(platformId, elementRef, inViewport) {
        this.platformId = platformId;
        this.elementRef = elementRef;
        this.inViewport = inViewport;
        this.config = new InViewportConfig();
        this.subscription = new Subscription();
        this.inViewportAction = new EventEmitter();
    }
    Object.defineProperty(InViewportDirective.prototype, "options", {
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.config = new InViewportConfig(value);
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    InViewportDirective.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        if (isPlatformBrowser(this.platformId)) {
            this.inViewport.register(this.elementRef.nativeElement, this.config);
            this.subscription.add(this.inViewport.trigger$
                .pipe(filter(function (entry) { return entry && entry.target === _this.elementRef.nativeElement; }))
                .subscribe(function (entry) { return _this.emitAction(entry, false); }));
        }
        else {
            this.emitAction(undefined, true);
        }
    };
    /**
     * @return {?}
     */
    InViewportDirective.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this.subscription.unsubscribe();
        if (isPlatformBrowser(this.platformId)) {
            this.inViewport.unregister(this.elementRef.nativeElement, this.config);
        }
    };
    /**
     * @param {?} entry
     * @param {?} force
     * @return {?}
     */
    InViewportDirective.prototype.check = /**
     * @param {?} entry
     * @param {?} force
     * @return {?}
     */
    function (entry, force) {
        var _this = this;
        var /** @type {?} */ isVisible = function () {
            var /** @type {?} */ partiallyVisible = entry.isIntersecting || entry.intersectionRatio > 0;
            var /** @type {?} */ completelyVisible = entry.intersectionRatio >= 1;
            return _this.config.partial ? partiallyVisible : completelyVisible;
        };
        var /** @type {?} */ visible = force || !entry || isVisible();
        return _a = {}, _a[InViewportMetadata] = { entry: entry }, _a.target = this.elementRef.nativeElement, _a.visible = visible, _a;
        var _a;
    };
    /**
     * @param {?} entry
     * @param {?} force
     * @return {?}
     */
    InViewportDirective.prototype.emitAction = /**
     * @param {?} entry
     * @param {?} force
     * @return {?}
     */
    function (entry, force) {
        var /** @type {?} */ event = this.config.checkFn
            ? this.config.checkFn(entry, { force: force, config: this.config })
            : this.check(entry, force);
        this.inViewportAction.emit(event);
    };
    InViewportDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[inViewport]'
                },] },
    ];
    /** @nocollapse */
    InViewportDirective.ctorParameters = function () { return [
        { type: Object, decorators: [{ type: Inject, args: [PLATFORM_ID,] },] },
        { type: ElementRef, },
        { type: InViewportService, },
    ]; };
    InViewportDirective.propDecorators = {
        "options": [{ type: Input, args: ['inViewportOptions',] },],
        "inViewportAction": [{ type: Output },],
    };
    return InViewportDirective;
}());
export { InViewportDirective };
function InViewportDirective_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    InViewportDirective.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    InViewportDirective.ctorParameters;
    /** @type {!Object<string,!Array<{type: !Function, args: (undefined|!Array<?>)}>>} */
    InViewportDirective.propDecorators;
    /** @type {?} */
    InViewportDirective.prototype.config;
    /** @type {?} */
    InViewportDirective.prototype.subscription;
    /** @type {?} */
    InViewportDirective.prototype.inViewportAction;
    /** @type {?} */
    InViewportDirective.prototype.platformId;
    /** @type {?} */
    InViewportDirective.prototype.elementRef;
    /** @type {?} */
    InViewportDirective.prototype.inViewport;
}
