// ng-in-viewport/esm5/lib/in-viewport.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Injectable, NgZone } from '@angular/core';
import { Subject } from 'rxjs';
import * as i0 from "@angular/core";
/**
 * @record
 */
export function InViewportRegistryEntry() { }
function InViewportRegistryEntry_tsickle_Closure_declarations() {
    /** @type {?} */
    InViewportRegistryEntry.prototype.root;
    /** @type {?} */
    InViewportRegistryEntry.prototype.targets;
    /** @type {?} */
    InViewportRegistryEntry.prototype.observer;
}
var InViewportService = /** @class */ (function () {
    function InViewportService(ngZone) {
        this.ngZone = ngZone;
        this.trigger$ = new Subject();
        this.registry = [];
    }
    /**
     * @param {?} entries
     * @return {?}
     */
    InViewportService.prototype.emitTrigger = /**
     * @param {?} entries
     * @return {?}
     */
    function (entries) {
        var _this = this;
        if (Array.isArray(entries) && entries.length) {
            entries.forEach(function (entry) { return _this.trigger$.next(entry); });
        }
    };
    /**
     * @param {?=} element
     * @return {?}
     */
    InViewportService.prototype.getRootElement = /**
     * @param {?=} element
     * @return {?}
     */
    function (element) {
        return element && element.nodeType === Node.ELEMENT_NODE ? element : undefined;
    };
    /**
     * @param {?} root
     * @return {?}
     */
    InViewportService.prototype.findEntry = /**
     * @param {?} root
     * @return {?}
     */
    function (root) {
        var _this = this;
        return this.registry.find(function (entry) { return entry.root === _this.getRootElement(root); });
    };
    /**
     * @param {?} target
     * @param {?} config
     * @return {?}
     */
    InViewportService.prototype.register = /**
     * @param {?} target
     * @param {?} config
     * @return {?}
     */
    function (target, config) {
        var _this = this;
        this.ngZone.runOutsideAngular(function () {
            var /** @type {?} */ foundedEntry = _this.findEntry(config.root);
            if (foundedEntry && !foundedEntry.targets.has(target)) {
                foundedEntry.targets.add(target);
                foundedEntry.observer.observe(target);
            }
            else {
                var /** @type {?} */ options = {
                    root: _this.getRootElement(config.root),
                    rootMargin: config.rootMargin,
                    threshold: config.threshold
                };
                var /** @type {?} */ entry = {
                    root: _this.getRootElement(config.root),
                    targets: new Set([target]),
                    observer: new IntersectionObserver(function (entries) { return _this.ngZone.run(function () { return _this.emitTrigger(entries); }); }, options)
                };
                entry.observer.observe(target);
                _this.registry = tslib_1.__spread(_this.registry, [entry]);
            }
        });
    };
    /**
     * @param {?} target
     * @param {?} config
     * @return {?}
     */
    InViewportService.prototype.unregister = /**
     * @param {?} target
     * @param {?} config
     * @return {?}
     */
    function (target, config) {
        var _this = this;
        this.ngZone.runOutsideAngular(function () {
            var /** @type {?} */ foundedEntry = _this.findEntry(config.root);
            if (foundedEntry) {
                var observer = foundedEntry.observer, targets = foundedEntry.targets;
                if (targets.has(target)) {
                    observer.unobserve(target);
                    targets.delete(target);
                }
                if (targets.size === 0) {
                    observer.disconnect();
                    _this.registry = _this.registry.filter(function (entry) { return entry !== foundedEntry; });
                }
            }
        });
    };
    InViewportService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] },
    ];
    /** @nocollapse */
    InViewportService.ctorParameters = function () { return [
        { type: NgZone, },
    ]; };
    /** @nocollapse */ InViewportService.ngInjectableDef = i0.defineInjectable({ factory: function InViewportService_Factory() { return new InViewportService(i0.inject(i0.NgZone)); }, token: InViewportService, providedIn: "root" });
    return InViewportService;
}());
export { InViewportService };
function InViewportService_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    InViewportService.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    InViewportService.ctorParameters;
    /** @type {?} */
    InViewportService.prototype.trigger$;
    /** @type {?} */
    InViewportService.prototype.registry;
    /** @type {?} */
    InViewportService.prototype.ngZone;
}
