// ng-in-viewport/esm5/lib/in-viewport-config.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import * as tslib_1 from "tslib";
/** @enum {string} */
var InViewportConfigDirection = {
    BOTH: 'both',
    VERTICAL: 'vertical',
    HORIZONTAL: 'horizontal',
};
export { InViewportConfigDirection };
/**
 * @record
 */
export function InViewportConfigCheckFnOptions() { }
function InViewportConfigCheckFnOptions_tsickle_Closure_declarations() {
    /** @type {?} */
    InViewportConfigCheckFnOptions.prototype.force;
    /** @type {?} */
    InViewportConfigCheckFnOptions.prototype.config;
}
/**
 * @record
 */
export function InViewportConfigOptions() { }
function InViewportConfigOptions_tsickle_Closure_declarations() {
    /** @type {?|undefined} */
    InViewportConfigOptions.prototype.root;
    /** @type {?|undefined} */
    InViewportConfigOptions.prototype.rootMargin;
    /** @type {?|undefined} */
    InViewportConfigOptions.prototype.threshold;
    /** @type {?|undefined} */
    InViewportConfigOptions.prototype.partial;
    /** @type {?|undefined} */
    InViewportConfigOptions.prototype.direction;
    /** @type {?|undefined} */
    InViewportConfigOptions.prototype.checkFn;
}
var /** @type {?} */ DEFAULT_THRESHOLD = [0, 1];
var InViewportConfig = /** @class */ (function () {
    function InViewportConfig(options) {
        var _this = this;
        this._rootMargin = '0px 0px 0px 0px';
        this._threshold = tslib_1.__spread(DEFAULT_THRESHOLD);
        this._partial = true;
        this._direction = InViewportConfigDirection.BOTH;
        if (Object.prototype.toString.call(options) === '[object Object]') {
            ['root', 'rootMargin', 'threshold', 'partial', 'direction', 'checkFn'].forEach(function (prop) {
                if (options.hasOwnProperty(prop)) {
                    _this[prop] = options[prop];
                }
            });
        }
    }
    Object.defineProperty(InViewportConfig.prototype, "root", {
        get: /**
         * @return {?}
         */
        function () {
            return this._root;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._root = value && value.nodeType === 1 ? value : undefined;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(InViewportConfig.prototype, "rootMargin", {
        get: /**
         * @return {?}
         */
        function () {
            return this._rootMargin;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._rootMargin = value && typeof value === 'string' ? value : '0px 0px 0px 0px';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(InViewportConfig.prototype, "threshold", {
        get: /**
         * @return {?}
         */
        function () {
            return this._threshold;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            var /** @type {?} */ threshold = [];
            var /** @type {?} */ isValidThreshold = function (val) { return typeof val === 'number' && val >= 0 && val <= 1; };
            if (isValidThreshold(value)) {
                threshold = [value];
            }
            else if (Array.isArray(value) && value.length) {
                threshold = value.filter(function (val) { return isValidThreshold(val); });
            }
            if (threshold.length === 0) {
                threshold = tslib_1.__spread(DEFAULT_THRESHOLD);
            }
            this._threshold = threshold;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(InViewportConfig.prototype, "partial", {
        get: /**
         * @return {?}
         */
        function () {
            return this._partial;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._partial = !!value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(InViewportConfig.prototype, "direction", {
        get: /**
         * @return {?}
         */
        function () {
            return this._direction;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            var /** @type {?} */ isValidValue = function (val) {
                return ([
                    InViewportConfigDirection.BOTH,
                    InViewportConfigDirection.HORIZONTAL,
                    InViewportConfigDirection.VERTICAL
                ].indexOf(val) >= 0);
            };
            this._direction = isValidValue(value) ? value : InViewportConfigDirection.BOTH;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(InViewportConfig.prototype, "checkFn", {
        get: /**
         * @return {?}
         */
        function () {
            return this._checkFn;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._checkFn = value;
        },
        enumerable: true,
        configurable: true
    });
    return InViewportConfig;
}());
export { InViewportConfig };
function InViewportConfig_tsickle_Closure_declarations() {
    /** @type {?} */
    InViewportConfig.prototype._root;
    /** @type {?} */
    InViewportConfig.prototype._rootMargin;
    /** @type {?} */
    InViewportConfig.prototype._threshold;
    /** @type {?} */
    InViewportConfig.prototype._partial;
    /** @type {?} */
    InViewportConfig.prototype._direction;
    /** @type {?} */
    InViewportConfig.prototype._checkFn;
}
