// @nicky-lenaers/ngx-scroll-tracker/scroll-tracker.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ScrollTrackerDirective } from './scroll-tracker.directive';
import { ScrollTrackerService } from './scroll-tracker.service';
var ScrollTrackerModule = (function () {
    function ScrollTrackerModule() {
    }
    /**
     * Guaranteed singletons for provided Services across App.
     *
     * @return 				An Angular Module with Providers
     */
    ScrollTrackerModule.forRoot = function () {
        return {
            ngModule: ScrollTrackerModule,
            providers: [
                ScrollTrackerService
            ]
        };
    };
    return ScrollTrackerModule;
}());
export { ScrollTrackerModule };
ScrollTrackerModule.decorators = [
    { type: NgModule, args: [{
                imports: [
                    CommonModule
                ],
                declarations: [
                    ScrollTrackerDirective
                ],
                exports: [
                    ScrollTrackerDirective
                ]
            },] },
];
/** @nocollapse */
ScrollTrackerModule.ctorParameters = function () { return []; };
