// ng-in-viewport/esm2015/lib/in-viewport.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Injectable, NgZone } from '@angular/core';
import { Subject } from 'rxjs';
import * as i0 from "@angular/core";
/**
 * @record
 */
export function InViewportRegistryEntry() { }
function InViewportRegistryEntry_tsickle_Closure_declarations() {
    /** @type {?} */
    InViewportRegistryEntry.prototype.root;
    /** @type {?} */
    InViewportRegistryEntry.prototype.targets;
    /** @type {?} */
    InViewportRegistryEntry.prototype.observer;
}
export class InViewportService {
    /**
     * @param {?} ngZone
     */
    constructor(ngZone) {
        this.ngZone = ngZone;
        this.trigger$ = new Subject();
        this.registry = [];
    }
    /**
     * @param {?} entries
     * @return {?}
     */
    emitTrigger(entries) {
        if (Array.isArray(entries) && entries.length) {
            entries.forEach((entry) => this.trigger$.next(entry));
        }
    }
    /**
     * @param {?=} element
     * @return {?}
     */
    getRootElement(element) {
        return element && element.nodeType === Node.ELEMENT_NODE ? element : undefined;
    }
    /**
     * @param {?} root
     * @return {?}
     */
    findEntry(root) {
        return this.registry.find((entry) => entry.root === this.getRootElement(root));
    }
    /**
     * @param {?} target
     * @param {?} config
     * @return {?}
     */
    register(target, config) {
        this.ngZone.runOutsideAngular(() => {
            const /** @type {?} */ foundedEntry = this.findEntry(config.root);
            if (foundedEntry && !foundedEntry.targets.has(target)) {
                foundedEntry.targets.add(target);
                foundedEntry.observer.observe(target);
            }
            else {
                const /** @type {?} */ options = {
                    root: this.getRootElement(config.root),
                    rootMargin: config.rootMargin,
                    threshold: config.threshold
                };
                const /** @type {?} */ entry = {
                    root: this.getRootElement(config.root),
                    targets: new Set([target]),
                    observer: new IntersectionObserver((entries) => this.ngZone.run(() => this.emitTrigger(entries)), options)
                };
                entry.observer.observe(target);
                this.registry = [...this.registry, entry];
            }
        });
    }
    /**
     * @param {?} target
     * @param {?} config
     * @return {?}
     */
    unregister(target, config) {
        this.ngZone.runOutsideAngular(() => {
            const /** @type {?} */ foundedEntry = this.findEntry(config.root);
            if (foundedEntry) {
                const { observer, targets } = foundedEntry;
                if (targets.has(target)) {
                    observer.unobserve(target);
                    targets.delete(target);
                }
                if (targets.size === 0) {
                    observer.disconnect();
                    this.registry = this.registry.filter((entry) => entry !== foundedEntry);
                }
            }
        });
    }
}
InViewportService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] },
];
/** @nocollapse */
InViewportService.ctorParameters = () => [
    { type: NgZone, },
];
/** @nocollapse */ InViewportService.ngInjectableDef = i0.defineInjectable({ factory: function InViewportService_Factory() { return new InViewportService(i0.inject(i0.NgZone)); }, token: InViewportService, providedIn: "root" });
function InViewportService_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    InViewportService.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    InViewportService.ctorParameters;
    /** @type {?} */
    InViewportService.prototype.trigger$;
    /** @type {?} */
    InViewportService.prototype.registry;
    /** @type {?} */
    InViewportService.prototype.ngZone;
}
