// ng-in-viewport/esm2015/lib/in-viewport-config.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/** @enum {string} */
const InViewportConfigDirection = {
    BOTH: 'both',
    VERTICAL: 'vertical',
    HORIZONTAL: 'horizontal',
};
export { InViewportConfigDirection };
/**
 * @record
 */
export function InViewportConfigCheckFnOptions() { }
function InViewportConfigCheckFnOptions_tsickle_Closure_declarations() {
    /** @type {?} */
    InViewportConfigCheckFnOptions.prototype.force;
    /** @type {?} */
    InViewportConfigCheckFnOptions.prototype.config;
}
/**
 * @record
 */
export function InViewportConfigOptions() { }
function InViewportConfigOptions_tsickle_Closure_declarations() {
    /** @type {?|undefined} */
    InViewportConfigOptions.prototype.root;
    /** @type {?|undefined} */
    InViewportConfigOptions.prototype.rootMargin;
    /** @type {?|undefined} */
    InViewportConfigOptions.prototype.threshold;
    /** @type {?|undefined} */
    InViewportConfigOptions.prototype.partial;
    /** @type {?|undefined} */
    InViewportConfigOptions.prototype.direction;
    /** @type {?|undefined} */
    InViewportConfigOptions.prototype.checkFn;
}
const /** @type {?} */ DEFAULT_THRESHOLD = [0, 1];
export class InViewportConfig {
    /**
     * @param {?=} options
     */
    constructor(options) {
        this._rootMargin = '0px 0px 0px 0px';
        this._threshold = [...DEFAULT_THRESHOLD];
        this._partial = true;
        this._direction = InViewportConfigDirection.BOTH;
        if (Object.prototype.toString.call(options) === '[object Object]') {
            ['root', 'rootMargin', 'threshold', 'partial', 'direction', 'checkFn'].forEach((prop) => {
                if (options.hasOwnProperty(prop)) {
                    this[prop] = options[prop];
                }
            });
        }
    }
    /**
     * @return {?}
     */
    get root() {
        return this._root;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set root(value) {
        this._root = value && value.nodeType === 1 ? value : undefined;
    }
    /**
     * @return {?}
     */
    get rootMargin() {
        return this._rootMargin;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set rootMargin(value) {
        this._rootMargin = value && typeof value === 'string' ? value : '0px 0px 0px 0px';
    }
    /**
     * @return {?}
     */
    get threshold() {
        return this._threshold;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set threshold(value) {
        let /** @type {?} */ threshold = [];
        const /** @type {?} */ isValidThreshold = (val) => typeof val === 'number' && val >= 0 && val <= 1;
        if (isValidThreshold(value)) {
            threshold = [value];
        }
        else if (Array.isArray(value) && value.length) {
            threshold = value.filter((val) => isValidThreshold(val));
        }
        if (threshold.length === 0) {
            threshold = [...DEFAULT_THRESHOLD];
        }
        this._threshold = threshold;
    }
    /**
     * @return {?}
     */
    get partial() {
        return this._partial;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set partial(value) {
        this._partial = !!value;
    }
    /**
     * @return {?}
     */
    get direction() {
        return this._direction;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set direction(value) {
        const /** @type {?} */ isValidValue = (val) => {
            return ([
                InViewportConfigDirection.BOTH,
                InViewportConfigDirection.HORIZONTAL,
                InViewportConfigDirection.VERTICAL
            ].indexOf(val) >= 0);
        };
        this._direction = isValidValue(value) ? value : InViewportConfigDirection.BOTH;
    }
    /**
     * @return {?}
     */
    get checkFn() {
        return this._checkFn;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set checkFn(value) {
        this._checkFn = value;
    }
}
function InViewportConfig_tsickle_Closure_declarations() {
    /** @type {?} */
    InViewportConfig.prototype._root;
    /** @type {?} */
    InViewportConfig.prototype._rootMargin;
    /** @type {?} */
    InViewportConfig.prototype._threshold;
    /** @type {?} */
    InViewportConfig.prototype._partial;
    /** @type {?} */
    InViewportConfig.prototype._direction;
    /** @type {?} */
    InViewportConfig.prototype._checkFn;
}
