// ng-in-viewport/esm2015/lib/in-viewport.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Directive, ElementRef, EventEmitter, Inject, Input, Output, PLATFORM_ID } from '@angular/core';
import { isPlatformBrowser } from '@angular/common';
import { Subscription } from 'rxjs';
import { filter } from 'rxjs/operators';
import { InViewportConfig } from './in-viewport-config';
import { InViewportService } from './in-viewport.service';
export const /** @type {?} */ InViewportMetadata = Symbol('InViewportMetadata');
export class InViewportDirective {
    /**
     * @param {?} platformId
     * @param {?} elementRef
     * @param {?} inViewport
     */
    constructor(platformId, elementRef, inViewport) {
        this.platformId = platformId;
        this.elementRef = elementRef;
        this.inViewport = inViewport;
        this.config = new InViewportConfig();
        this.subscription = new Subscription();
        this.inViewportAction = new EventEmitter();
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set options(value) {
        this.config = new InViewportConfig(value);
    }
    /**
     * @return {?}
     */
    ngAfterViewInit() {
        if (isPlatformBrowser(this.platformId)) {
            this.inViewport.register(this.elementRef.nativeElement, this.config);
            this.subscription.add(this.inViewport.trigger$
                .pipe(filter((entry) => entry && entry.target === this.elementRef.nativeElement))
                .subscribe((entry) => this.emitAction(entry, false)));
        }
        else {
            this.emitAction(undefined, true);
        }
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        this.subscription.unsubscribe();
        if (isPlatformBrowser(this.platformId)) {
            this.inViewport.unregister(this.elementRef.nativeElement, this.config);
        }
    }
    /**
     * @param {?} entry
     * @param {?} force
     * @return {?}
     */
    check(entry, force) {
        const /** @type {?} */ isVisible = () => {
            const /** @type {?} */ partiallyVisible = entry.isIntersecting || entry.intersectionRatio > 0;
            const /** @type {?} */ completelyVisible = entry.intersectionRatio >= 1;
            return this.config.partial ? partiallyVisible : completelyVisible;
        };
        const /** @type {?} */ visible = force || !entry || isVisible();
        return { [InViewportMetadata]: { entry }, target: this.elementRef.nativeElement, visible };
    }
    /**
     * @param {?} entry
     * @param {?} force
     * @return {?}
     */
    emitAction(entry, force) {
        const /** @type {?} */ event = this.config.checkFn
            ? this.config.checkFn(entry, { force, config: this.config })
            : this.check(entry, force);
        this.inViewportAction.emit(event);
    }
}
InViewportDirective.decorators = [
    { type: Directive, args: [{
                selector: '[inViewport]'
            },] },
];
/** @nocollapse */
InViewportDirective.ctorParameters = () => [
    { type: Object, decorators: [{ type: Inject, args: [PLATFORM_ID,] },] },
    { type: ElementRef, },
    { type: InViewportService, },
];
InViewportDirective.propDecorators = {
    "options": [{ type: Input, args: ['inViewportOptions',] },],
    "inViewportAction": [{ type: Output },],
};
function InViewportDirective_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    InViewportDirective.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    InViewportDirective.ctorParameters;
    /** @type {!Object<string,!Array<{type: !Function, args: (undefined|!Array<?>)}>>} */
    InViewportDirective.propDecorators;
    /** @type {?} */
    InViewportDirective.prototype.config;
    /** @type {?} */
    InViewportDirective.prototype.subscription;
    /** @type {?} */
    InViewportDirective.prototype.inViewportAction;
    /** @type {?} */
    InViewportDirective.prototype.platformId;
    /** @type {?} */
    InViewportDirective.prototype.elementRef;
    /** @type {?} */
    InViewportDirective.prototype.inViewport;
}
